File: src/0139-Word-Break/0139.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution
{
public:
    bool wordBreak(string s, vector<string>& wordDict)
    {
        vector<int> mem(s.size() + 1, 0);
        mem[0] = true;
        for (unsigned int i = 1; i < s.size() + 1; ++i)
        {
            for (auto& word : wordDict)
            {
                if (i >= word.size() && mem[i - word.size()] &&
                    word == s.substr(i - word.size(), word.size()))
                {
                    mem[i] = true;
                    break;
                }
            }
        }
        return *mem.rbegin();
    }
};

File: src/0139-Word-Break/0139.py
class Solution:
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        return self._wordBreak(s, set(wordDict), 0, set())
        
    def _wordBreak(self, s, words, start, mem):
        if start == len(s):
            return True
        
        if start in mem:
            return False

        for i in range(start + 1, len(s) + 1):
            if i in mem:
                continue

            sub = s[start:i]
            if sub in words and self._wordBreak(s, words, i, mem):
                return True

        mem.add(start)
        return False

if __name__ == '__main__':
    s = "applepenapple"
    wordDict = ["apple", "pen"]
    print(Solution().wordBreak(s, wordDict))